devteam-402/cmput402-w71
# Set up a boilerplate for backend development using Docker

**Purpose:** Set up a boilerplate for backend development using Docker
**Tests:** Successfully set it up on both OS X and Ubuntu (Cybera http://162.246.157.148/api/admin)
**Test Instructions:**
In order to build and run api-server on your machine properly, you need to go through the following steps.

For Mac users:
1. Make sure your Docker and VirtualBox has been installed properly

2. Create a new virtual machine in VirtualBox:
```docker-machine create -d virtualbox default```

3. You can run this virtual machine by typing
```make env```
 and when you see machine "default" is running, run the following command to configure your shell.
```eval $(docker-machine env default)```
Now, the credentials for connecting "default" virtual machine are all set in current shell. Please keep in mind that those credentials are only available temporarily in the shell that you just configured. Once the the shell is closed, those credentials are no longer exist. In other words, whenever you opened a new shell, you need to re-configure it using the two commands above.

4. Build images for the services that are defined in docker-compose.yml
```make build```

5. Run all services
```make run```

6. Test your web service with ```http://your_docker_machine_ip/api/admin``` in your browser and you should be able to see the Django admin page! Also, you can try to log into admin dashboard as a root user. (Username: root Password: root) 
There are several ways to find out the IP address of your docker machine. One way is to use command ```docker-machine ls``` to list the status of your docker machine and another way is to run ```make env``` in a seperate tab.




For Ubuntu/other linux distribution users:
1. Make sure you have installed docker-engine and recommended linux kernels correctly by following this tutorial https://docs.docker.com/engine/installation/linux/ubuntu/

2. Install docker-compose by following this tutorial
https://docs.docker.com/compose/install/
Note: Do NOT use ```apt-get``` to install any docker related tools, because a lot of those binaries are not up-to-date!

4. Build images for the services that are defined in docker-compose.yml
```make build```

5. Run all services
```make run```

**Core files:**

**Makefile** - provide a set of commands for building and running Docker images on your local machine.
**docker-compose.yml** - mainly specify which Docker images will be used and how different containers link to each other in a development context.
**circletest.yml** - similar to docker-compose but specifically made for CircleCI's virtual machines.
**production.yml** - similar to docker-compose but specifically made for production environment (Cybera).
***/Dockerfile** - contains all the commands a user could call on the command line to assemble a new Docker image.
